File: src/ios/TCPlugin.m
//
//  TCPlugin.h
//  Twilio Client plugin for PhoneGap / Cordova
//
//  Copyright 2012 Stevie Graham.
//


#import "TCPlugin.h"
#import  <AVFoundation/AVFoundation.h>


@interface TCPlugin() {
    TCDevice     *_device;
    TCConnection *_connection;
    NSString     *_callback;
}

@property(nonatomic, strong) TCDevice     *device;
@property(nonatomic, strong) NSString     *callback;
@property(atomic, strong)    TCConnection *connection;
@property(atomic, strong)    UILocalNotification *ringNotification;

-(void)javascriptCallback:(NSString *)event;
-(void)javascriptCallback:(NSString *)event withArguments:(NSDictionary *)arguments;
-(void)javascriptErrorback:(NSError *)error;

@end

@implementation TCPlugin

@synthesize device     = _device;
@synthesize callback   = _callback;
@synthesize connection = _connection;
@synthesize ringNotification = _ringNotification;

# pragma mark device delegate method

-(void)device:(TCDevice *)device didStopListeningForIncomingConnections:(NSError *)error {
    [self javascriptErrorback:error];
}

-(void)device:(TCDevice *)device didReceiveIncomingConnection:(TCConnection *)connection {
    self.connection = connection;    
    [self javascriptCallback:@"onincoming"];
}

-(void)device:(TCDevice *)device didReceivePresenceUpdate:(TCPresenceEvent *)presenceEvent {
    NSString *available = [NSString stringWithFormat:@"%d", presenceEvent.isAvailable];
    NSDictionary *object = [NSDictionary dictionaryWithObjectsAndKeys:presenceEvent.name, @"from", available, @"available", nil];
    [self javascriptCallback:@"onpresence" withArguments:object];
}

-(void)deviceDidStartListeningForIncomingConnections:(TCDevice *)device {
    // What to do here? The JS library doesn't have an event for this.
}

# pragma mark connection delegate methods

-(void)connection:(TCConnection*)connection didFailWithError:(NSError*)error {
    [self javascriptErrorback:error];
}

-(void)connectionDidStartConnecting:(TCConnection*)connection {
    self.connection = connection;
    // What to do here? The JS library doesn't have an event for connection negotiation.
}

-(void)connectionDidConnect:(TCConnection*)connection {
    self.connection = connection;
    [self javascriptCallback:@"onconnect"];
    if([connection isIncoming]) [self javascriptCallback:@"onaccept"];
}

-(void)connectionDidDisconnect:(TCConnection*)connection {
    self.connection = connection;
    [self javascriptCallback:@"ondevicedisconnect"];
    [self javascriptCallback:@"onconnectiondisconnect"];
}

# pragma mark javascript device mapper methods

-(void)deviceSetup:(CDVInvokedUrlCommand*)command {
    self.callback = command.callbackId;
    self.device = [[TCDevice alloc] initWithCapabilityToken:[command.arguments objectAtIndex:0] delegate:self];
    
    // Disable sounds. was getting EXC_BAD_ACCESS
    //self.device.incomingSoundEnabled   = NO;
    //self.device.outgoingSoundEnabled   = NO;
    //self.device.disconnectSoundEnabled = NO;

    [NSTimer scheduledTimerWithTimeInterval:1 target:self selector:@selector(deviceStatusEvent) userInfo:nil repeats:NO];
}

-(void)deviceStatusEvent {
    switch ([self.device state]) {
        case TCDeviceStateReady:
            [self javascriptCallback:@"onready"];
            NSLog(@"State: Ready");
            break;
            
        case TCDeviceStateOffline:
            [self javascriptCallback:@"onoffline"];
            NSLog(@"State: Offline");
            break;
            
        default:
            break;
    }
}

-(void)connect:(CDVInvokedUrlCommand*)command {
    [self.device connect:[command.arguments objectAtIndex:0] delegate:self];
}

-(void)disconnectAll:(CDVInvokedUrlCommand*)command {
    [self.device disconnectAll];
}

-(void)deviceStatus:(CDVInvokedUrlCommand*)command {
    NSString *state;
    switch ([self.device state]) {
        case TCDeviceStateBusy:
            state = @"busy";
            break;
            
        case TCDeviceStateReady:
            state = @"ready";
            break;
            
        case TCDeviceStateOffline:
            state = @"offline";
            break;
            
        default:
            break;        
    }
    
    CDVPluginResult *result = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK messageAsString:state];    
    [self performSelectorOnMainThread:@selector(writeJavascript:) withObject:[result toSuccessCallbackString:command.callbackId] waitUntilDone:NO];
}


# pragma mark javascript connection mapper methods

-(void)acceptConnection:(CDVInvokedUrlCommand*)command {
    [self.connection accept];
}

-(void)disconnectConnection:(CDVInvokedUrlCommand*)command {
    [self.connection disconnect];
}

-(void)rejectConnection:(CDVInvokedUrlCommand*)command {
    [self.connection reject];
}

-(void)muteConnection:(CDVInvokedUrlCommand*)command {
    if(self.connection.isMuted) {
        self.connection.muted = NO;
    } else {
        self.connection.muted = YES;
    }
}

-(void)sendDigits:(CDVInvokedUrlCommand*)command {
    [self.connection sendDigits:[command.arguments objectAtIndex:0]];
}

-(void)connectionStatus:(CDVInvokedUrlCommand*)command {
    NSString *state;
    switch ([self.connection state]) {
        case TCConnectionStateConnected:
            state = @"open";
            break;
            
        case TCConnectionStateConnecting:
            state = @"connecting";
            break;
            
        case TCConnectionStatePending:
            state = @"pending";
            break;
            
        case TCConnectionStateDisconnected:
            state = @"closed";
        
        default:
            break;        
    }
        
    CDVPluginResult *result = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK messageAsString:state];    
    [self performSelectorOnMainThread:@selector(writeJavascript:) withObject:[result toSuccessCallbackString:command.callbackId] waitUntilDone:NO];
}

-(void)connectionParameters:(CDVInvokedUrlCommand*)command {
    CDVPluginResult *result = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK messageAsDictionary:[self.connection parameters]];
    [self performSelectorOnMainThread:@selector(writeJavascript:) withObject:[result toSuccessCallbackString:command.callbackId] waitUntilDone:NO];
}


-(void)showNotification:(CDVInvokedUrlCommand*)command {
    @try {
        [[UIApplication sharedApplication] cancelAllLocalNotifications];
    }
    @catch(NSException *exception) {
        NSLog(@"Couldn't Cancel Notification");
    }
    
    NSString *alertBody = [command.arguments objectAtIndex:0];
    
    NSString *ringSound = @"incoming.wav";
    if([command.arguments count] == 2) {
        ringSound = [command.arguments objectAtIndex:1];
    }

    _ringNotification = [[UILocalNotification alloc] init];
    _ringNotification.alertBody = alertBody;
    _ringNotification.alertAction = @"Answer";
    _ringNotification.soundName = ringSound;
    _ringNotification.fireDate = [NSDate date];
    [[UIApplication sharedApplication] scheduleLocalNotification:_ringNotification];

}

-(void)cancelNotification:(CDVInvokedUrlCommand*)command {
    [[UIApplication sharedApplication] cancelLocalNotification:_ringNotification];
}

-(void)setSpeaker:(CDVInvokedUrlCommand*)command {
    NSString *mode = [command.arguments objectAtIndex:0];
    if([mode isEqual: @"on"]) {
        UInt32 audioRouteOverride = kAudioSessionOverrideAudioRoute_Speaker;
        AudioSessionSetProperty (
            kAudioSessionProperty_OverrideAudioRoute,
            sizeof (audioRouteOverride),
            &audioRouteOverride
        );
    }
    else {
        UInt32 audioRouteOverride = kAudioSessionOverrideAudioRoute_None;
        AudioSessionSetProperty (
            kAudioSessionProperty_OverrideAudioRoute,
            sizeof (audioRouteOverride),
            &audioRouteOverride
        );
    }
}

# pragma mark private methods

-(void)javascriptCallback:(NSString *)event withArguments:(NSDictionary *)arguments {
    NSDictionary *options   = [NSDictionary dictionaryWithObjectsAndKeys:event, @"callback", arguments, @"arguments", nil];
    CDVPluginResult *result = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK messageAsDictionary:options];
    result.keepCallback     = [NSNumber numberWithBool:YES];
    
    [self performSelectorOnMainThread:@selector(writeJavascript:) withObject:[result toSuccessCallbackString:self.callback] waitUntilDone:NO];
}

-(void)javascriptCallback:(NSString *)event {
    [self javascriptCallback:event withArguments:nil];
}

-(void)javascriptErrorback:(NSError *)error {
    NSDictionary *object    = [NSDictionary dictionaryWithObjectsAndKeys:[error localizedDescription], @"message", nil];
    CDVPluginResult *result = [CDVPluginResult resultWithStatus:CDVCommandStatus_ERROR messageAsDictionary:object];
    result.keepCallback     = [NSNumber numberWithBool:YES];
    
    [self performSelectorOnMainThread:@selector(writeJavascript:) withObject:[result toErrorCallbackString:self.callback] waitUntilDone:NO];
}

@end

File: www/tcPlugin.js
(function() {
    var delegate = {}
    var TwilioPlugin = {
        Device: function() {
            return this;
        },

        Connection: function() {
            return this;
        }
    }


    TwilioPlugin.Device.prototype.setup = function(token) {
        // Take a token and instantiate a new device object
        var error = function(error) {
            if(delegate['ondeviceerror']) delegate['ondeviceerror'](error)
            if(delegate['onconnectionerror']) delegate['onconnectionerror'](error)
        }

        var success = function(callback) {
            var argument = callback['arguments'] || new TwilioPlugin.Connection();
            if (delegate[callback['callback']]) delegate[callback['callback']](argument);
        }

        Cordova.exec(success, error, "TCPlugin", "deviceSetup", [token]);
    }

    // polymorphic function. if called with function as an argument, the function is invoked
    // when a connection has been established. if called with an object a connection is established with those options
    TwilioPlugin.Device.prototype.connect = function(argument) {
        if (typeof(argument) == 'function') {
            delegate['onconnect'] = argument;
        } else if (typeof(argument) == 'object') {
            Cordova.exec(null,null,"TCPlugin","connect", [argument])
        }
    }

    TwilioPlugin.Device.prototype.disconnectAll = function() {
        Cordova.exec(null,null,"TCPlugin","disconnectAll",[]);
    }

    TwilioPlugin.Device.prototype.disconnect = function(fn) {
        delegate['ondevicedisconnect'] = fn;
    }

    TwilioPlugin.Device.prototype.ready = function(fn) {
        delegate['onready'] = fn;
    }

    TwilioPlugin.Device.prototype.offline = function(fn) {
        delegate['onoffline'] = fn;
    }

    TwilioPlugin.Device.prototype.incoming = function(fn) {
        delegate['onincoming'] = fn;
    }

    TwilioPlugin.Device.prototype.cancel = function(fn) {
        delegate['oncancel'] = fn;
    }

    TwilioPlugin.Device.prototype.error = function(fn) {
        delegate['ondeviceerror'] = fn;
    }

    TwilioPlugin.Device.prototype.presence = function(fn) {
        delegate['onpresence'] = fn;
    }

    TwilioPlugin.Device.prototype.status = function(fn) {
        Cordova.exec(fn,null,"TCPlugin","deviceStatus",[]);
    }

    TwilioPlugin.Device.prototype.sounds = {
        incoming: function(boolean) {},
        outgoing: function(boolean) {},
        disconnect: function(boolean) {}
    }

    TwilioPlugin.Connection.prototype.accept = function(argument) {
        if (typeof(argument) == 'function') {
            delegate['onaccept'] = argument;
        } else {
            Cordova.exec(null,null,"TCPlugin","acceptConnection",[]);
        }
    }

    TwilioPlugin.Connection.prototype.showNotification = function(alertBody, ringSound) {
        var args = [alertBody, ringSound];
        if(ringSound === "undefined") {
            args = [alertBody];
        }    
        Cordova.exec(null, null, "TCPlugin", "showNotification", args);
    }

    TwilioPlugin.Connection.prototype.cancelNotification = function() {
        Cordova.exec(null, null, "TCPlugin", "cancelNotification", []);
    }

    TwilioPlugin.Connection.prototype.setSpeaker = function(mode) {
        // "on" or "off"        
        Cordova.exec(null, null, "TCPlugin", "setSpeaker", [mode]);
    }

    TwilioPlugin.Connection.prototype.reject = function() {
        Cordova.exec(null,null,"TCPlugin","rejectConnection",[]);
    }

    TwilioPlugin.Connection.prototype.disconnect = function(fn) {
        if (typeof(argument) == 'function') {
            delegate['onconnectiondisconnect'] = argument;
        } else {
            Cordova.exec(null,null,"TCPlugin","disconnectConnection",[]);
        }
    }

    TwilioPlugin.Connection.prototype.error = function(fn) {
        delegate['onconnectionerror'] = fn;
    }

    TwilioPlugin.Connection.prototype.mute = function() {
        Cordova.exec(null,null,"TCPlugin","muteConnection",[]);
    }

    TwilioPlugin.Connection.prototype.unmute = function() {
        Cordova.exec(null,null,"TCPlugin","muteConnection",[]);
    }

    TwilioPlugin.Connection.prototype.sendDigits = function(string) {
        Cordova.exec(null,null,"TCPlugin","sendDigits", [string]);
    }

    TwilioPlugin.Connection.prototype.status = function(fn) {
        Cordova.exec(fn, null, "TCPlugin", "connectionStatus", []);
    }
	
	    TwilioPlugin.Connection.prototype.parameters = function(fn) {
        Cordova.exec(fn, null, "TCPlugin", "connectionParameters", []);
    }

    TwilioPlugin.install = function() {
        if (!window.Twilio) window.Twilio = {};
        if (!window.Twilio.Device) window.Twilio.Device = new TwilioPlugin.Device();
        if (!window.Twilio.Connection) window.Twilio.Connection = new TwilioPlugin.Connection();
    }
 TwilioPlugin.install();

})()